### PR TITLE
[WIP][doc] Tutorial: How to add a MC move

### PR DESCRIPTION
**Pull request by g-bauer**
This PR adds a section to the "advanced tutorials" of the user documentation about "how to add a move to the MC propagator".

### Structure and work packages
I created the tutorial with the following structure in mind (which are also the work packages and the doc sections):

- [x] Overview and introduction.
- [x] Add a "naive" version of a molecular translation move (reimplement the current move).
- [ ] Use the move in a NVT simulations of an atomic liquid. Talk about efficiency, both in execution time and accepted moves. 

This will be used as motivation to talk about:
- [ ] Increase efficiency of "naive" approach by using the energy cache and adjustment of the displacement.

Finally,
- [ ] Expose the move to the input crate. Perform the same simulation as above again using an input file, comparing the efficiency + summary.

Up to now, I pick up MC and rust specific stuff on the fly. The structure aims at incrementally raising the difficulty, both on the rust and the MC side of things.

### Feedback please!
I'd love to hear some feedback on what is currently implemented. Don't hesitate to give feedback on content/format/wording and style/code examples/structure etc. Also, what do you think about the degree of complexity? Is it too easy or too complicated? Is the style of writing OK or too "casual"?

Thanks in advance!

### CODE REVIEW

**Comment by Luthaf** [comment]
The outline and organisation looks good. I'll read the tutorial later!

The biggest question here will be which level of Rust knowledge should we assume from the users? One possibility is to refer to the official Rust book everytime we introduce a new Rust concept; another one is to write our own explanations in the stream of the tutorial. The second way might make it easier to get started, but I am afraid we end up rewriting the Rust book =/

**Comment by g-bauer** [comment]
Currently, I have a list at the start of the the overview section about what one needs to know for the tutorial. I might move it to the subsections, since it is very different for each section. 

At the moment, I only assume the reader is familiar with what a `struct` is, how `function`s work and what the point of a `Trait` is. As I write this, I notice that I'll have to add `Vectors`. To be honest, I'd favor to keep the rust specific part as low as possible (of course, it has to be specific to the point that the user understands how to use the APIs etc.). Linking to the book would be my take.

In the end, it is a tutorial for Lumol, not rust 😄

**Comment by antoinewdg** [comment]
This is pretty good !

Some remarks:
* I am 100% in favor of linking to the sections of the Rust book containing required knowledge for each section, rewriting the Rust book would be a loss of time
* why do we store the whole system ? I know that it is purposefully not an optimal implementation, but wouldn't storing only the old energy and the sampled displacement be more natural ? 
* it feels a little weird having a tutorial explaining how to implement MC move with some very noob-friendly explanations of what we are doing when we do not have a noob-friendly explanation of what MC simulation is
* the whole part explaining how to use particles shows the need of a tutorial for the system (where this part would belong)
* I really like the "let's make the obvious but inefficient one first and iterate" approach, it clearly helps understanding 

> Is it too easy or too complicated? 

In my opinion you can never make a tutorial (or any educational content for that matter) too easy, as a contributor your are so heavily biased by your knowledge of the internals that it's really hard to grasp what a user may or may not know and to estimate the sheer amount of new information one has to assimilate in order to understand things well. Plus, even if you make it too easy, it's still infinitely better than making too hard :smile: .

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
+# A minimalist implementation
+
+> What is a MC move again?  
+> In MC, moves are used to propagate the system.
+> That is, a move is a specific instruction about how to change the current configuration and to propose a new configuration.
+> Configuration here means the positions of all atoms (and not the velocities - in contrast to molecular dynamics).
+> A move can be accepted - then the system will adopt the proposed configuration - or rejected - then the system will not change.
+> Moves can never happen simultaneously - they will take place always one after another
```
Comment (g-bauer) [inline doc/src/advanced_tutorial/add_move_minimal.md]: That is a good point, thanks.

**Comment by g-bauer** [comment]
Thanks a lot for to you both for your feedback, very much appreciated!

> why do we store the whole system ?

As you mention, it is on purpose because it is the simplest way to do it, since we have a strict separation between `preparation` (where we have a mutable system) and `cost` (immutable ref to system where all computation should take place).  I use this example as a naive way to implement the move. If it is too far-fetched to be plausible, I think we can come up with something different.

> it feels a little weird having a tutorial explaining how to implement MC move with some very noob-friendly explanations of what we are doing when we do not have a noob-friendly explanation of what MC simulation is

Mhm, the question is then: where to start? I assume that people that are interested in extending the code with a new move, have an understanding of how MC works. But as you say, do these people need the  "noob-friendly" explanations we currently have in the first place? Also, there is room for explanations about the inner workings of MC and MD in the conceptual section of the manual. So it does not have to be in form of a tutorial.  

> the whole part explaining how to use particles shows the need of a tutorial for the system (where this part would belong)

That is true. Maybe it is also a hint for us that the API could be improved?

> Plus, even if you make it too easy, it's still infinitely better than making too hard 😄 .

Also true. I'll go with adding as much information as I think is needed. In the end, we can cut excess stuff easily.

**Comment by Luthaf** [comment]
>  Another idea would be to also teach the reader how to use the code docs and link to the API documentation when needed?

We should definitively do this ! As I see it, the goal is to help users to go from simple users => plugins developers => contributors => core dev! The plugin part will need #45, to allow people to add there own code without touching the core of the simulation engine.

File: doc/src/advanced_tutorial/add_move.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+└── moves
+    ├── mod.rs
+    ├── rotate.rs
+    :
+    └── translate.rs
+```
+
+Initially, all of our work will take place within the `moves` directory.
+
+In rust terms, a move is a struct that implements the `MCMove` trait. The trait is defined in the `moves/mod.rs` file. Let's have a look:
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move.md]: Could we have a way to make sure these examples stay up-to-date ?

File: doc/src/advanced_tutorial/add_move.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    fn restore(&mut self, system: &mut System);
+    fn update_amplitude(&mut self, scaling_factor: Option<f64>);
+}
+```
+
+In Lumol, MC propagation (which is defined in `monte-carlo.rs`) is divided into several steps.
+Basically, each of these steps makes use of one of the above shown methods of `MCMove`.
+As you can see the `System` structure is involved in most functions.
+The `System` is the heart of Lumol and is a quite sophisticated structure.
+At the moment it is sufficient for us to notice that the system contains the positions (`system.positions`) and that we can access the energy of the current configuration directly (`system.total_energy()`).
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move.md]: `system.positions` does not exist

File: doc/src/advanced_tutorial/add_move.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+Going through the functions one-by-one:
+
+- The `describe` function returns a `&str` with a brief description of the move, e.g. "translation of a molecule" or "rotation of a molecule". Primarily, this function is used for log files and errors.
+- `setup` is called *once before the actual simulation is run* for every move. You can use this function to set `System` dependent parameter.
+- `prepare` can be used to perform small computations to set up the moves' parameter that depend on the state of the `System`.
+- The `cost` function is the core functionality of a move. It is the function where most (if not all) of the computational work takes place. The return value is used from the propagator to decide if a proposed move is accepted or rejected. 
+- If a move is accepted, we update the `System` by invoking the `apply` function.
+- If a move is rejected, we restore the `System` to its old state by invoking the `restore` function (if we modified it beforehand).
+- The `update_amplitude` function can be used to increase the efficiency of a simulation run by adjusting how much a move actually changes the system.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move.md]: Maybe adding a note / link to an explanation for this, and saying not to worry about it in the simplest cases?

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
+# A minimalist implementation
+
+> What is a MC move again?  
+> In MC, moves are used to propagate the system.
+> That is, a move is a specific instruction about how to change the current configuration and to propose a new configuration.
+> Configuration here means the positions of all atoms (and not the velocities - in contrast to molecular dynamics).
+> A move can be accepted - then the system will adopt the proposed configuration - or rejected - then the system will not change.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: Maybe talk/link to the Metropolis criterion here?

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+```rust
+/// Monte-Carlo move for translating a molecule
+pub struct Translate {
+    /// System state prior to the move
+    old_system: System,
+    /// Maximum displacement value
+    dr: f64,
+    /// Translation range for random number generation
+    range: Range<f64>,
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: The `Range<f64>` looks advanced to me. How about just using `rng.gen_range(-self.dr, self.dr)` in the code?

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            old_system: System::default(),
+            dr: dr,
+            range: Range::new(-dr, dr),
+        }
+    }
+}
+```
+
+The `new` function takes a single argument, the maximum displacement `dr`.
+The internal unit of length - the displacement is a length - is Angstrom.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: We don't need to specify the internal unit. We can just say that all lengths are guaranteed to be in internal units, and maybe link to the internal unit definition.

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+```rust
+impl Default for Translate {
+    fn default() -> Translate {
+        Translate::new(1.0)
+    }
+}
+``` 
+
+We also implement `Default`, which simply calls `new` with a displacement of `dr = 1.0 angstrom`.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: We don't actually need this here. The Default implementation are mainly here to satisfy clippy, but I believe they are not used. We might want to remove them.

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        system[i].position += delta
+    }
+    true // we set everything up, so we can perform the move.
+}
+```
+
+That is a lot of stuff.
+Let's first have a look at the function arguments:
+We take three mutable references: One to the `Translate` struct (`&mut self`), one
+to the system and finally one to `Box<Rng>`, which is a **r**andom **n**umber **g**enerator.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: I would write this `Random Number Generator`

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+Let us now go through the code blocks one by one.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: You could add markers `(1)`, `(2)`, `(3)` in the comments and refer to them here.

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+First, we randomly select a molecule.
+All molecules are stored in a vector inside system, `system.molecules`.
+We cannot access this vector directly, since it is not public, but there is a public function that we can use: `system.molecules()` will return the vector of molecules.
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: We don't need to explain the internal structure of a `System`. The idea of an API is to encapsulate the internals.

I would write "We can access the list of molecules in the system by calling the system.molecules() function".

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+In Lumol, positions are stored as `Vector3D` structures.
+It is a vector with three elements, each of which stands for a cartesian coordinate.
+We create a random displacement vector by drawing a random number for every cartesian coordinate.
+In rust, we can do that by taking a sample from `range` which we set to be `[-dr, dr]` in our `new()` function.
+
+To apply the displacement, we first store the current state of the system using the `clone` function.
+In the `for` loop, we add the displacement by modifying the positions of the system.
+To understand what's going on here, we have to take a closer look on how positions are stored in Lumol and how we can access them:
+```rust
+for i in self.old_system.molecule(molid) {
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: How about `for particle_index in system.molecule(molecule_index)`? It feels clearer to me.

File: doc/src/advanced_tutorial/add_move_minimal.md [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+Finally, we can access the position of a particle with index  `i` using `system[i].position`.
+
+The cost function is comparably short but don't get fooled as it is the most expensive function of the move.
+
+```rust
+fn cost(&self, system: &System, beta: f64, _: &mut EnergyCache) -> f64 {
+    let new_energy = self.old_system.total_energy();
+    let old_energy = system.total_energy();
+    let delta_energy = new_energy - old_energy;
+    beta * delta_energy // return the cost function
```
Comment (Luthaf) [inline doc/src/advanced_tutorial/add_move_minimal.md]: // return the cost

there is no function here.

**Comment by Luthaf** [comment]
So the exact wording looks pretty good to me.

The only hard part is the one with the index and molecules and everything. I feels like this part is trying to explain too many things at once. I would rather go for the minimal explanations to get what is going on, not explaining how things are stored or why we put particles contiguous in memory and the bonds in a separated struct and so on.

Instead, I think we can appeal to the intuition of the user: we are iterating over the particles contained in a molecule. For efficiency reasons, we are not storing the particles directly in the molecules, but the particles indexes. With the index, we can access the particles in the system, and the position of the particle.

**Comment by antoinewdg** [comment]
## Comparing to master (e817acf6)
using `--threshold 2, latest commit first`
<details><summary>95685007 Added naive implementation section.</summary>
  <p>





```bash




```



</p></details>

<details><summary>75d9d748 Added tutorial on how to implement a MC move.</summary>
  <p>





```bash




```



</p></details>

## Individual benchmarks

<details><summary>95685007 Added naive implementation section.</summary>
  <p>





```bash


running 4 tests
test cache_move_particle ... bench:     242,300 ns/iter (+/- 2,138)
test energy              ... bench:   1,847,067 ns/iter (+/- 1,963)
test forces              ... bench:   2,319,834 ns/iter (+/- 2,880)
test virial              ... bench:   2,716,298 ns/iter (+/- 1,706)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test cache_move_particle_ewald ... bench:     428,242 ns/iter (+/- 773)
test cache_move_particle_wolf  ... bench:     106,252 ns/iter (+/- 194)
test energy_ewald              ... bench:     887,737 ns/iter (+/- 2,044)
test energy_wolf               ... bench:     494,427 ns/iter (+/- 907)
test forces_ewald              ... bench:  15,817,525 ns/iter (+/- 55,027)
test forces_wolf               ... bench:     670,223 ns/iter (+/- 927)
test virial_ewald              ... bench:  47,831,527 ns/iter (+/- 39,029)
test virial_wolf               ... bench:     662,513 ns/iter (+/- 1,591)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test cache_move_all_rigid_molecules ... bench:   1,040,967 ns/iter (+/- 23,761)
test cache_move_particles           ... bench:     350,248 ns/iter (+/- 2,151)
test energy                         ... bench:   1,122,353 ns/iter (+/- 1,361)
test forces                         ... bench:   1,371,558 ns/iter (+/- 1,840)
test virial                         ... bench:   1,442,552 ns/iter (+/- 20,721)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test cache_move_all_rigid_molecules_ewald ... bench:   8,764,994 ns/iter (+/- 61,997)
test cache_move_all_rigid_molecules_wolf  ... bench:   8,368,377 ns/iter (+/- 32,945)
test cache_move_particles_ewald           ... bench:     900,016 ns/iter (+/- 2,066)
test cache_move_particles_wolf            ... bench:     494,391 ns/iter (+/- 875)
test energy_ewald                         ... bench:   1,573,232 ns/iter (+/- 2,507)
test energy_wolf                          ... bench:     741,049 ns/iter (+/- 1,282)
test forces_ewald                         ... bench:  23,264,315 ns/iter (+/- 27,276)
test forces_wolf                          ... bench:   1,064,350 ns/iter (+/- 1,139)
test virial_ewald                         ... bench:  69,525,596 ns/iter (+/- 749,333)
test virial_wolf                          ... bench:   1,009,285 ns/iter (+/- 1,610)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<details><summary>75d9d748 Added tutorial on how to implement a MC move.</summary>
  <p>





```bash


running 4 tests
test cache_move_particle ... bench:     243,028 ns/iter (+/- 2,171)
test energy              ... bench:   1,845,139 ns/iter (+/- 4,845)
test forces              ... bench:   2,322,307 ns/iter (+/- 4,404)
test virial              ... bench:   2,722,362 ns/iter (+/- 2,836)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test cache_move_particle_ewald ... bench:     428,499 ns/iter (+/- 906)
test cache_move_particle_wolf  ... bench:     106,497 ns/iter (+/- 368)
test energy_ewald              ... bench:     903,793 ns/iter (+/- 2,356)
test energy_wolf               ... bench:     495,901 ns/iter (+/- 3,257)
test forces_ewald              ... bench:  16,662,926 ns/iter (+/- 28,243)
test forces_wolf               ... bench:     670,037 ns/iter (+/- 1,556)
test virial_ewald              ... bench:  47,837,361 ns/iter (+/- 425,185)
test virial_wolf               ... bench:     661,155 ns/iter (+/- 1,274)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test cache_move_all_rigid_molecules ... bench:   1,039,589 ns/iter (+/- 22,377)
test cache_move_particles           ... bench:     349,818 ns/iter (+/- 1,483)
test energy                         ... bench:   1,119,416 ns/iter (+/- 11,641)
test forces                         ... bench:   1,370,352 ns/iter (+/- 66,873)
test virial                         ... bench:   1,440,662 ns/iter (+/- 1,136)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test cache_move_all_rigid_molecules_ewald ... bench:   8,570,057 ns/iter (+/- 42,202)
test cache_move_all_rigid_molecules_wolf  ... bench:   8,176,199 ns/iter (+/- 11,516)
test cache_move_particles_ewald           ... bench:     888,121 ns/iter (+/- 1,626)
test cache_move_particles_wolf            ... bench:     484,861 ns/iter (+/- 1,596)
test energy_ewald                         ... bench:   1,572,268 ns/iter (+/- 3,353)
test energy_wolf                          ... bench:     825,654 ns/iter (+/- 4,346)
test forces_ewald                         ... bench:  23,234,240 ns/iter (+/- 25,132)
test forces_wolf                          ... bench:   1,014,101 ns/iter (+/- 1,382)
test virial_ewald                         ... bench:  71,534,166 ns/iter (+/- 418,388)
test virial_wolf                          ... bench:     995,827 ns/iter (+/- 2,864)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<details><summary>e817acf6 Change the ordering of the arguments</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     243,341 ns/iter (+/- 3,504)
test argon::energy              ... bench:   1,846,614 ns/iter (+/- 9,595)
test argon::forces              ... bench:   2,357,906 ns/iter (+/- 3,469)
test argon::virial              ... bench:   2,712,546 ns/iter (+/- 4,410)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     411,205 ns/iter (+/- 1,745)
test nacl::cache_move_particle_wolf  ... bench:     106,293 ns/iter (+/- 232)
test nacl::energy_ewald              ... bench:     864,591 ns/iter (+/- 3,330)
test nacl::energy_wolf               ... bench:     471,234 ns/iter (+/- 528)
test nacl::forces_ewald              ... bench:  17,549,466 ns/iter (+/- 59,919)
test nacl::forces_wolf               ... bench:     663,047 ns/iter (+/- 796)
test nacl::virial_ewald              ... bench:  48,120,993 ns/iter (+/- 106,867)
test nacl::virial_wolf               ... bench:     662,466 ns/iter (+/- 753)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   1,036,638 ns/iter (+/- 27,723)
test propane::cache_move_particles           ... bench:     351,038 ns/iter (+/- 4,754)
test propane::energy                         ... bench:   1,126,473 ns/iter (+/- 13,071)
test propane::forces                         ... bench:   1,380,765 ns/iter (+/- 1,649)
test propane::virial                         ... bench:   1,441,503 ns/iter (+/- 1,256)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:   8,427,446 ns/iter (+/- 17,482)
test water::cache_move_all_rigid_molecules_wolf  ... bench:   8,159,764 ns/iter (+/- 27,635)
test water::cache_move_particles_ewald           ... bench:     859,472 ns/iter (+/- 2,267)
test water::cache_move_particles_wolf            ... bench:     482,360 ns/iter (+/- 1,488)
test water::energy_ewald                         ... bench:   1,539,201 ns/iter (+/- 2,464)
test water::energy_wolf                          ... bench:     733,296 ns/iter (+/- 517)
test water::forces_ewald                         ... bench:  25,706,464 ns/iter (+/- 38,325)
test water::forces_wolf                          ... bench:     996,237 ns/iter (+/- 2,432)
test water::virial_ewald                         ... bench:  69,794,941 ns/iter (+/- 417,351)
test water::virial_wolf                          ... bench:     987,888 ns/iter (+/- 3,265)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<br>**CPU**: Intel(R) Core(TM) i7-5700HQ CPU @ 2.70GHz

**Pull request closed without merging**